homeassistant/components/strava/__init__.py

```diff
@@ -0,0 +1,242 @@
+        self._client_secret = config.get(CONF_CLIENT_SECRET)
+
+        self.athletes = {}
+        self.gears = {}
+        self.clubs = {}
+
+    @property
+    def is_authorized(self):
+        return self._token is not None
+
+    @property
+    def is_token_valid(self):
+        if not self.is_authorized:
+            _LOGGER.info("Not authorized")
+            return False
+
+        expires_at = datetime.fromtimestamp(self._token['expires_at'])
+        if expires_at > datetime.now() + 300:
+            return True
+
+        _LOGGER.info("Token expired: %s", repr(self._token))
+        return False
+
+    async def get_token(self):
+        if not self.is_authorized:
+            store = self._hass.helpers.storage.Store(STORAGE_VERSION,
+                                                     STORAGE_KEY)
+            self._token = await store.async_load()
+
+            if self._token:
+                self.client.access_token = self._token['access_token']
+            else:
+                _LOGGER.info("Requesting token")
+                await self.request_token()
+                return
+
+        expires_at = datetime.fromtimestamp(self._token['expires_at'])
+        if expires_at < datetime.now():
+            await self.refresh_token()
+
+    async def authorize(self, code, hass):
+        """ Request initial authorization. """
+        self._token = await hass.async_add_executor_job(
+            self.client.exchange_code_for_token,
+            self._client_id,
+            self._client_secret,
+            code
+        )
+
+        store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+        await store.async_save(self._token)
+
+        if self.is_authorized:
+            await hass.async_add_executor_job(
+                hass.components.configurator.request_done,
+                self._configurator
+            )
+            del self._configurator
+
+        await async_setup(hass, self._config)
+
+    async def request_token(self):
+        """Request Strava access token."""
+
+        callback_url = '{}{}'.format(self._hass.config.api.base_url,
+                                     AUTH_CALLBACK_PATH)
+        authorize_url = self.client.authorization_url(
+            client_id=self._config.get(CONF_CLIENT_ID),
+            redirect_uri=callback_url)
+
+        self._configurator = \
+            self._hass.components.configurator.async_request_config(
+                DEFAULT_NAME, lambda _: None,
+                link_name=CONFIGURATOR_LINK_NAME,
+                link_url=authorize_url,
+                description=CONFIGURATOR_DESCRIPTION,
+                submit_caption=CONFIGURATOR_SUBMIT_CAPTION)
+
+    async def refresh_token(self):
+        """Renew Strava access token."""
+
+        self._token = await self._hass.async_add_executor_job(
+            self.client.refresh_access_token,
+            self._client_id,
+            self._client_secret,
+            self._token['refresh_token'])
+
```

Comment:
Should the client token be updated here?

` self.client.access_token = self._token['access_token']
`

Reply:
The `client.refresh_access_token()` function in stravalib, already does this:
https://github.com/hozn/stravalib/blob/03abab18b8f0b3a7ed371ff0f109374fea8a9750/stravalib/protocol.py#L163

We just need to save the token to our store.